File: tests/test_integration.py
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from hp_guard.parser import PolicyParser
from hp_guard.models import PolicyCall


POLICY = """
version: 1
global:
  default_action: allow

agents:
  research-bot:
    tools:
      curl:
        rules:
          - action: deny
            condition:
              args_match: ".*--delete.*"
      write_file:
        rules:
          - action: deny
            condition:
              path_pattern: "/etc/*"
"""


def test_research_bot_curl_delete_denied():
    engine = PolicyParser.parse(POLICY)
    call = PolicyCall(agent="research-bot", tool="curl", args=["--delete", "http://x.com"])
    decision = engine.resolve_call(call)
    assert decision.action.value == "deny"


def test_research_bot_curl_get_allowed():
    engine = PolicyParser.parse(POLICY)
    call = PolicyCall(agent="research-bot", tool="curl", args=["--get", "http://x.com"])
    decision = engine.resolve_call(call)
    assert decision.action.value == "allow"


def test_research_bot_write_file_etc_denied():
    engine = PolicyParser.parse(POLICY)
    call = PolicyCall(agent="research-bot", tool="write_file", args=["/etc/passwd"])
    decision = engine.resolve_call(call)
    assert decision.action.value == "deny"


def test_unknown_agent_defaults_to_allow():
    engine = PolicyParser.parse(POLICY)
    call = PolicyCall(agent="unknown-agent", tool="anything")
    decision = engine.resolve_call(call)
    assert decision.action.value == "allow"

File: tests/test_integration.rs
//! Full end-to-end integration tests mirroring the Python integration suite.
use hp_guard::models::PolicyCall;
use hp_guard::parser::PolicyParser;
use hp_guard::Action;

const POLICY: &str = r#"
version: 1
global:
  default_action: allow

agents:
  research-bot:
    tools:
      curl:
        rules:
          - action: deny
            condition:
              args_match: ".*--delete.*"
      write_file:
        rules:
          - action: deny
            condition:
              path_pattern: "/etc/*"
"#;

#[test]
fn test_research_bot_curl_delete_denied() {
    let engine = PolicyParser::parse(POLICY).expect("policy is valid");
    let call = PolicyCall {
        agent: Some("research-bot".into()),
        tool: Some("curl".into()),
        args: vec!["--delete".into(), "http://x.com".into()],
        user: None,
        context: Default::default(),
    };
    let decision = engine.resolve_call(&call);
    assert_eq!(decision.action, Action::Deny);
}

#[test]
fn test_research_bot_curl_get_allowed() {
    let engine = PolicyParser::parse(POLICY).expect("policy is valid");
    let call = PolicyCall {
        agent: Some("research-bot".into()),
        tool: Some("curl".into()),
        args: vec!["--get".into(), "http://x.com".into()],
        user: None,
        context: Default::default(),
    };
    let decision = engine.resolve_call(&call);
    assert_eq!(decision.action, Action::Allow);
}

#[test]
fn test_research_bot_write_file_etc_denied() {
    let engine = PolicyParser::parse(POLICY).expect("policy is valid");
    let call = PolicyCall {
        agent: Some("research-bot".into()),
        tool: Some("write_file".into()),
        args: vec!["/etc/passwd".into()],
        user: None,
        context: Default::default(),
    };
    let decision = engine.resolve_call(&call);
    assert_eq!(decision.action, Action::Deny);
}

#[test]
fn test_unknown_agent_defaults_to_allow() {
    let engine = PolicyParser::parse(POLICY).expect("policy is valid");
    let call = PolicyCall {
        agent: Some("unknown-agent".into()),
        tool: Some("anything".into()),
        args: Default::default(),
        user: None,
        context: Default::default(),
    };
    let decision = engine.resolve_call(&call);
    assert_eq!(decision.action, Action::Allow);
}
